### PR TITLE
Fe 1313 button group

### PR DESCRIPTION
**Pull request by dquilter**
Updates the `ButtonToggle` and `ButtonToggleGroup` components:

- Removes Input Decorator
- Adds styles for DLS
- Moves styling from SASS to Styled Components

Classic Theme `ButtonToggle` x3:

Screenshot 2019-04-25 at 10 32 35

Modern Theme `ButtonToggleGroup` with focused `ButtonToggle`:

Screenshot 2019-04-25 at 10 16 38

### CODE REVIEW

**Comment by benwilsonsage** [comment]
Looking great - some things that might be already in progress - I know it's an early draft...
- 'Classic' theme looks like 'Small' right now.
- Adding 12px of margin/spacing between buttons would be perfect.
- 3px gold outline for a focus state.
- The border colour for the selected state should be the same colour as the background.

File: src/components/button-toggle/button-toggle.style.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+const StyledButtonToogle = styled.div`
+  display: inline;
+`;
+
+const StyledButtonToogleLabel = styled.label`
+  display: inline-block;
+  height: 40px;
+  padding: 0 24px;
+  border: 1px solid ${({ theme }) => theme.colors.border};
+  font-size: 14px;
```
Comment (izakrzewska) [inline src/components/button-toggle/button-toggle.style.js]: maybe `theme.text.size`?

File: src/components/button-toggle/button-toggle.style.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      line-height: 40px;
+    }
+    .carbon-icon {
+      margin-bottom: 8px;
+    }
+  `}
+
+  ${({ theme }) => theme.name === THEMES.classic && css`
+    .carbon-icon::before {
+      font-size: 60px;
```
Comment (izakrzewska) [inline src/components/button-toggle/button-toggle.style.js]: Zrzut ekranu 2019-04-18 o 09 19 00
why is icon size 60px?

File: src/components/button-toggle/button-toggle.style.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  font-size: 14px;
+  font-weight: 600;
+
+  input:checked ~ & {
+    color: ${({ theme }) => theme.colors.white};
+    background-color: ${({ theme }) => theme.colors.tertiary};
+  }
+
+  &:hover {
+    background-color: ${({ theme }) => theme.colors.whiteMix};
```
Comment (izakrzewska) [inline src/components/button-toggle/button-toggle.style.js]: what about background color styling for classic?

File: src/style/themes/base/base-theme.config.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ export default (palette) => {
       secondary: palette.genericGreenShade(35),
       tertiary: palette.genericGreenShade(55),
       brand: palette.brilliantGreen,
+      whiteMix: '#e5f2ee',
```
Comment (izakrzewska) [inline src/style/themes/base/base-theme.config.js]: is this color hover color for all themes?

File: src/components/button-toggle/button-toggle.component.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Icon from '../icon';
+import { StyledButtonToogle, StyledButtonToogleLabel, StyledButtonToggleIcon } from './button-toggle.style';
+import guid from '../../utils/helpers/guid';
+
+const ButtonToggle = (props) => {
+  const inputGuid = guid();
+
+  function icon() {
```
Comment (grabkowski) [inline src/components/button-toggle/button-toggle.component.js]: could you change that name to be more meaningful? e.g.: `renderIcon`.

File: src/components/button-toggle/button-toggle.component.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    if (!props.buttonIcon) return null;
+
+    return (
+      <StyledButtonToggleIcon { ...props }>
+        <Icon type={ props.buttonIcon } />
+      </StyledButtonToggleIcon>
+    );
+  }
+
+  function hiddenInput() {
```
Comment (grabkowski) [inline src/components/button-toggle/button-toggle.component.js]: maybe `renderHiddenInput`?

File: src/components/button-toggle/button-toggle.component.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    return (
+      <input
+        type='radio'
+        name={ props.name }
+        id={ inputGuid }
+        onChange={ onChangeHandler }
+        style={ { // ToDo: Move to styled-component?
```
Comment (grabkowski) [inline src/components/button-toggle/button-toggle.component.js]: we will not need hidden inputs at all

File: src/components/button-toggle/button-toggle.component.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  function onChangeHandler(ev) {
+    props.onChange(ev, this.props);
+  }
+
+  return (
+    <StyledButtonToogle>
+      {hiddenInput()}
+      <StyledButtonToogleLabel { ...props } htmlFor={ inputGuid }>
+        <div className='content-wrapper'>
```
Comment (grabkowski) [inline src/components/button-toggle/button-toggle.component.js]: that should be a styled-component

File: src/components/button-toggle/button-toggle.style.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    outline: 3px solid ${({ theme }) => theme.colors.focus};
+  }
+
+  .content-wrapper {
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    width: 100%;
+    height: 100%;
+  }
```
Comment (grabkowski) [inline src/components/button-toggle/button-toggle.style.js]: that should be a style-component

**Comment by grabkowski** [comment]
I think that should be a Draft PR instead of regular one

File: src/components/button-toggle/button-toggle.component.js (hunk 5 above)
Comment (dquilter) [reply src/components/button-toggle/button-toggle.component.js]: I think it's fairly clear what this method is doing.

File: src/components/button-toggle/button-toggle.component.js (hunk 6 above)
Comment (dquilter) [reply src/components/button-toggle/button-toggle.component.js]: I think it's fairly clear what this method is doing.

File: src/components/button-toggle/button-toggle.style.js (hunk 3 above)
Comment (dquilter) [reply src/components/button-toggle/button-toggle.style.js]: The `classic` theme should never use `whiteMix`, the colour is a new part of the DLS.

File: src/components/button-toggle/button-toggle.style.js (hunk 9 above)
Comment (dquilter) [reply src/components/button-toggle/button-toggle.style.js]: Resolving as being discussed elsewhere: https://github.com/Sage/carbon/pull/2021/files#r276644913

**Comment by benwilsonsage** [comment]
Just one minor UX thing... in small/medium/large themes, the buttons are slightly indented on the left... could we line up the label, button, and help? Maybe just removing margin on the first child perhaps.
Screenshot 2019-04-25 at 08 36 07

File: src/components/button-toggle/button-toggle.component.js (hunk 8 above)
Comment (grabkowski) [reply src/components/button-toggle/button-toggle.component.js]: I think that we shouldn't mix up CSS with styled components like that

File: src/components/button-toggle-group/button-toggle-group.stories.js [hunk 10]
```diff
@@ -1,10 +1,12 @@
...
 import { storiesOf } from '@storybook/react';
-import { text, boolean, select } from '@storybook/addon-knobs';
+import {
+  text, number, boolean, select, percentageRange
+} from '@storybook/addon-knobs';
 import OptionsHelper from '../../utils/helpers/options-helper';
-import notes from './notes.md';
-import ButtonToggle from '../button-toggle/button-toggle';
-import ButtonToggleGroup from './button-toggle-group';
+import notes from './documentation/notes.md';
```
Comment (dawidzarzycki) [inline src/components/button-toggle-group/button-toggle-group.stories.js]: Can we have index.js file in the documentation and export the notes?

File: src/components/button-toggle/button-toggle.stories.js [hunk 11]
```diff
@@ -3,50 +3,62 @@ import { storiesOf } from '@storybook/react';
 import {
   text, select, boolean
 } from '@storybook/addon-knobs';
+import { action } from '@storybook/addon-actions';
 import OptionsHelper from '../../utils/helpers/options-helper';
-import notes from './notes.md';
-import ButtonToggle from './button-toggle';
+import notes from './documentation/notes.md';
```
Comment (dawidzarzycki) [inline src/components/button-toggle/button-toggle.stories.js]: Can we have index.js file in the documentation and export the notes?

File: src/components/button-toggle/button-toggle.style.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    }
+  `}
+
+  ${({ disabled }) => disabled && css`
+    background-color: #e6ebed !important;
+    border-color: #e6ebed !important;
+    color: rgba(0,0,0,.2) !important;
+  `};
+
+  /* CLASSIC STYLES */
```
Comment (dawidzarzycki) [inline src/components/button-toggle/button-toggle.style.js]: I think it should be in a separate file eg `classic-button-toggle.style.js`

File: src/components/button-toggle/package.json [hunk 13]
```diff
@@ -1,3 +1,3 @@
 {
```
Comment (dawidzarzycki) [inline src/components/button-toggle/package.json]: Can we replace package.json with index.js?

File: src/style/themes/base/base-theme.config.js (hunk 4 above)
Comment (dawidzarzycki) [reply src/style/themes/base/base-theme.config.js]: I'm not sure about the colors in themes that they are not created by palette. @samhinton88 Could you take a look at this? I had similar problem before but I don't remember exactly what was that about

File: src/components/button-toggle/button-toggle.spec.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+          <ButtonToggle { ...props } />
+          <ButtonToggle { ...props } />
+        </div>
+      );
+      // Uses snapshot as jest/enzyme doesnt support :first-of-type
+      expect(wrapper).toMatchSnapshot();
+    });
+  });
+  describe('As a form element', () => {
+    it('calls onChange when the value changes', () => {
```
Comment (stevebutler2210) [inline src/components/button-toggle/button-toggle.spec.js]: Is this test necessary?

File: src/components/button-toggle/button-toggle.style.js [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+    height: 104px;
+    padding: 0 16px;
+
+    ${StyledButtonToggleContentWrapper} {
+      flex-direction: column;
+    }
+  `}
+
+  ${({ disabled }) => disabled && css`
+    background-color: #e6ebed !important;
```
Comment (stevebutler2210) [inline src/components/button-toggle/button-toggle.style.js]: Can this go in a theme variable?

File: src/components/button-toggle/package.json (hunk 13 above)
Comment (dquilter) [reply src/components/button-toggle/package.json]: I thought the `package.json` was needed for the demo site? So it should get deleted when that gets trashed?

File: src/components/button-toggle/button-toggle.spec.js (hunk 14 above)
Comment (dquilter) [reply src/components/button-toggle/button-toggle.spec.js]: Removed

File: src/components/button-toggle-group/documentation/index.js [hunk 16]
```diff
@@ -0,0 +1,3 @@
+import notes from './notes.md';
```
Comment (dawidzarzycki) [inline src/components/button-toggle-group/documentation/index.js]: Can we export in one line?  `export {  default } from './notes.md'`

File: src/components/button-toggle/package.json (hunk 13 above)
Comment (dquilter) [reply src/components/button-toggle/package.json]: Updated

File: src/components/button-toggle/button-toggle.component.js [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+  StyledButtonToggleInput,
+  StyledButtonToggleIcon,
+  StyledButtonToggleContentWrapper
+} from './button-toggle.style';
+import guid from '../../utils/helpers/guid';
+
+const ButtonToggle = (props) => {
+  const inputGuid = guid();
+
+  function icon() {
```
Comment (samhinton88) [inline src/components/button-toggle/button-toggle.component.js]: What is the benefit in declaring these functions inside the body of the `ButtonToggle` assignment? I would consider moving where you compose these to either outside of the main component, or preferably into their own file.

File: src/components/button-toggle/button-toggle-icon.component.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { StyledButtonToggleIcon } from './button-toggle.style';
+import Icon from '../icon';
+
+function ButtonToggleIcon(props) {
+  if (!props.buttonIcon) return null;
```
Comment (stevebutler2210) [inline src/components/button-toggle/button-toggle-icon.component.js]: Could we move this check up a level, into button-toggle.component.js, to conditionally render the component?

File: src/components/button-toggle/button-toggle.style.js [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+  align-items: center;
+  width: 100%;
+  height: 100%;
+`;
+
+const StyledButtonToggleLabel = styled.label`
+  display: inline-block;
+  height: 38px;
+  padding: 0 24px;
+  border: 1px solid ${({ theme }) => theme.colors.border};
```
Comment (stevebutler2210) [inline src/components/button-toggle/button-toggle.style.js]: Could you wrap the styles in an arrow function to pull in `theme` at the top level? Not too precious about it normally, but there's a lot of repetition going on here

File: src/components/button-toggle-group/__snapshots__/button-toggle-group.spec.js.snap [hunk 20]
```diff
@@ -0,0 +1,429 @@
...
+          },
+        }
+      }
+      value="bar"
+    >
+      Bar
+    </ButtonToggle>
+  </styled.div>
+</FormField>
+`;
```
Comment (grabkowski) [inline src/components/button-toggle-group/__snapshots__/button-toggle-group.spec.js.snap]: more than 90% of this snapshot is a theme object. Could you change the render method to `shallow` or use `TestRenderer.create` if you're checking styles?

File: src/components/button-toggle-group/button-toggle-group.spec.js [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+  describe('Classic theme', () => {
+    it('renders correctly with default settings', () => {
+      const wrapper = render({ theme: classicTheme });
+      expect(wrapper).toMatchSnapshot();
+    });
+  });
+  describe('Modern theme', () => {
+    it('renders correctly with default settings', () => {
+      const wrapper = render({ theme: smallTheme });
+      expect(wrapper).toMatchSnapshot();
```
Comment (grabkowski) [inline src/components/button-toggle-group/button-toggle-group.spec.js]: snapshots generated by these tests differs only with the theme prop. Please use `TestRenderer.create` to properly render css classes output from `styled-components`.

File: src/components/button-toggle/__snapshots__/button-toggle.spec.js.snap [hunk 22]
```diff
@@ -0,0 +1,1266 @@
...
+    size="large"
+  >
+    <div
+      className="c3 c4"
+    >
+      Button
+    </div>
+  </label>
+</div>
+`;
```
Comment (grabkowski) [inline src/components/button-toggle/__snapshots__/button-toggle.spec.js.snap]: another snapshot composed mostly of theme objects. Please use `TestRenderer.create` to render it properly.

File: src/components/button-toggle/button-toggle-classic.style.js [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+    font-weight: 700;
+    font-size: 12px;
+  `};
+
+  ${({ size, buttonIcon, buttonIconSize }) => buttonIcon && size === 'large' && buttonIconSize === 'large' && css`
+    height: auto;
+    padding-top: 15px;
+    padding-bottom: 15px;
+
+    .carbon-button-toggle__button-icon {
```
Comment (grabkowski) [inline src/components/button-toggle/button-toggle-classic.style.js]: isn't that a separate component now?

Comment (dquilter) [reply src/components/button-toggle/button-toggle-classic.style.js]: Yes, but this is styling specific to the `Classic` theme. Based on our current approach to `Classic` I don't think it's worth wasting time changing this.

**Pull request merged**